### PR TITLE
Add GOST 34.11 hash algorithm

### PR DESCRIPTION
**Pull request by dmitrykatenko**
Add OTP generation using GOST 34.11 hash function with 256/512 bit length (gost.ts, key-utilities.ts). Add necessary options and enumerations (AddAccountPge.vue, otp.ts).

### CODE REVIEW

File: src/models/key-utilities.ts [hunk 1]
```diff
@@ -91,6 +92,28 @@ export class KeyUtilities {
...
   }
 
+  static CryptoJsWordArrayToUint8Array(wordArray: CryptoJS.lib.WordArray) {
+    const l = wordArray.sigBytes;
+    const words = wordArray.words;
+    const result = new Uint8Array(l);
+    let i = 0 /*dst*/,
+      j = 0; /*src*/
+    // eslint-disable-next-line no-constant-condition
+    while (true) {
```
Comment (Sneezry) [inline src/models/key-utilities.ts]: Seems `while(i < l)` makes more sense here?

File: src/models/key-utilities.ts [hunk 2]
```diff
@@ -91,6 +92,28 @@ export class KeyUtilities {
...
+      // here i is a multiple of 4
+      if (i == l) break;
```
Comment (Sneezry) [inline src/models/key-utilities.ts]: Should always use `===` instead of `==`

File: src/models/key-utilities.ts (hunk 1 above)
Comment (dmitrykatenko) [reply src/models/key-utilities.ts]: Agree with you. Corrected

File: src/models/key-utilities.ts (hunk 2 above)
Comment (dmitrykatenko) [reply src/models/key-utilities.ts]: OK, made a commit

**Review by Sneezry — APPROVED** [review]
LGTM

**Review by mymindstorm — COMMENTED** [review]
My preliminary concerns before looking at the code too closely / testing:
- I have not heard of this algorithm before, is it used for TOTP? It appears to be nonstandard as it is not referenced in RFC 6238
- Due to the above I would like to see automated tests on this. It will be harder to maintain nonstandard algorithms from our end, so tests would be very beneficial. I will be able to help with this if needed when I have some free time.
- It looks like you took code from another one of your company's projects to implement GOST. Is this code available as a library on npm?

**Comment by dmitrykatenko** [comment]
- the totp algorithm allows a variety of hash functions to be used on the client and provider, including GOST. Since the hash function is also an input parameter of the totp algorithm. http://tools.ietf.org/html/rfc6986 - RFC to hash function.
- ok, we will work on it
- it is not in the npm repository, but it's probably not a problem to place

**Comment by dmitrykatenko** [comment]
Implemented npm package usage

**Comment by dmitrykatenko** [comment]
tests added

**Review by mymindstorm — CHANGES_REQUESTED** [review]
Although the test you have added is good to verify that the library is functioning, what I am looking for is code that preforms a test with known good HOTP secrets and codes against codes generated with GOST from `KeyUtilites.generate`. I want to make sure that any future changes to our code does not impact GOST since we will not be testing it manually. After that we can merge this.

**Comment by dmitrykatenko** [comment]
made a test for predefined secret and date with known/valid otp code

**Review by mymindstorm — APPROVED** [review]

**Comment by mymindstorm** [comment]
Thanks for the changes!

**Comment by dmitrykatenko** [comment]
please tell me when the release is planned with our changes?

**Comment by mymindstorm** [comment]
We don't have anything planned, but I have been looking to fix #676 before the next release.

**Pull request merged**